packages/core/core/src/requests/BundleGraphRequest.js

```diff
@@ -145,23 +145,24 @@ class BundlerRunner {
       }
     }
 
-    // let cacheKey;
-    // if (
-    //   !this.options.shouldDisableCache// &&
-    //   // !this.api.hasInvalidRequests()
-    // ) {
     let cacheKey = await this.getCacheKey(graph, configResult);
-    //   let cachedBundleGraphBuffer;
-    //   try {
-    //     cachedBundleGraphBuffer = await this.options.cache.getBlob(cacheKey);
-    //   } catch {
-    //     // Cache miss
-    //   }
-
-    //   if (cachedBundleGraphBuffer) {
-    //     return [deserialize(cachedBundleGraphBuffer), cachedBundleGraphBuffer];
-    //   }
-    // }
+
+    // Check if the cacheKey matches the one already stored in the graph.
+    // This can save time deserializing from cache if the graph is already in memory.
+    let previousResult = await this.api.getPreviousResult(cacheKey);
+    if (previousResult != null) {
+      this.api.storeResult(previousResult, cacheKey);
+      return previousResult;
+    }
+
+    // Otherwise, check the cache in case the cache key has already been written to disk.
+    if (!this.options.shouldDisableCache) {
+      let cached = await this.options.cache.get(cacheKey);
+      if (cached != null) {
+        this.api.storeResult(cached, cacheKey);
+        return cached;
+      }
+    }
```

Comment:
@wbinnssmith what do you think we should do here about serializing and deserializing the graph? Currently this would deserialize the graph from cache, and immediately re-serialize it when sending to the workers. And again serialize it on process exit when writing to disk.

1. Should we store the graph in memory (in the request graph) at all? That can save time deserializing from cache in cases where a file changes that invalidates the bundle graph request (e.g. package.json), but the cache key doesn't actually change (e.g. changing a different key in package.json than the bundle graph config). But it may also use a bunch of memory. At least there should only be one instance of the bundle graph around at a time though.
2. Should we store the original serialized buffer as part of the BundleGraph itself when reading from cache so that when we need to send it to workers and write to disk later we don't need to serialize it again? Again the tradeoff is storing the serialized buffer in memory, which I think you were concerned about in #5300.

Reply:
1. We should probably retain the BundleGraph in memory after packaging, as we'll likely need to do this for incremental bundling anyway. cc @joeyslater @AGawrys 
2. At one point I believe we were serializing the graph multiple times in succession to send it to each worker. We should make sure to only do so once when we package. Other than that I don't know if we need to keep it around. 
3. Yeah this is basically a cpu/memory tradeoff when shutting down, i.e.: is it better to shut down quickly or to use less memory? In watch mode I'd probably prefer the latter.